lib/src/revset.rs

```diff
@@ -4327,4 +4336,29 @@ mod tests {
         }
         "###);
     }
+
+    #[test]
+    fn test_escape_string_literal() {
+        // Valid identifiers don't need quoting
+        assert_eq!(format_symbol("foo"), "foo");
+        assert_eq!(format_symbol("foo.bar"), "foo.bar");
+
+        // Invalid identifiers need quoting
+        assert_eq!(format_symbol("foo@bar"), r#""foo@bar""#);
+        assert_eq!(format_symbol("foo bar"), r#""foo bar""#);
+        assert_eq!(format_symbol("(foo)"), r#""(foo)""#);
+        assert_eq!(format_symbol("all:foo"), r#""all:foo""#);
+
+        // Some characters also need escaping
+        assert_eq!(format_symbol(r#"foo"bar"#), r#""foo\"bar""#);
+        assert_eq!(format_symbol(r#"foo\bar"#), r#""foo\\bar""#);
+        assert_eq!(format_symbol(r#"foo\"bar"#), r#""foo\\\"bar""#);
+        assert_eq!(format_symbol("foo\nbar"), r#""foo\nbar""#);
+
+        // Some characters don't technically need escaping, but we escape them for
+        // clarity
+        assert_eq!(format_symbol(r#"foo"bar"#), r#""foo\"bar""#);
+        assert_eq!(format_symbol(r#"foo\bar"#), r#""foo\\bar""#);
+        assert_eq!(format_symbol(r#"foo\"bar"#), r#""foo\\\"bar""#);
+    }
```

Comment:
nit: can you add test for symbol starts/ends with space? It could be an edge case if our grammar had implicit whitespace rule.